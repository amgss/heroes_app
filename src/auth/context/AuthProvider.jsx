import { useReducer } from 'react';
import { types } from '../types/types';
import { AuthContext } from './AuthContext';
import { authReducer } from './authReducer';

const initialState = { logged: false };
const init = () => {
    const user = JSON.parse(localStorage.getItem('user'));

    return {
        logged: !!user,
        user
    };
};

export const AuthProvider = ({ children }) => {
    const [authState, dispatch] = useReducer(authReducer, initialState, init);
    const login = (name = '') => {
        const user = { id: 1, name };

        const action = {
            type: types.login,
            payload: user
        };

        localStorage.setItem('user', JSON.stringify(user));

        dispatch(action);
    };

    const logout = (name = '') => {
        localStorage.removeItem('user');
        const action = {
            type: types.logout,
            payload: {
                id: null,
                name: ''
            }
        };

        dispatch(action);
    };

    return (
        <AuthContext.Provider value={{ ...authState, login, logout }}>
            {children}
        </AuthContext.Provider>
    )
}
