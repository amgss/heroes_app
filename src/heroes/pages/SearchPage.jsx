import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';
import { HeroCard } from '../components';
import { getHeroesByName } from '../helpers'
import { useForm } from '../../hooks/FormCustomHook';

export const SearchPage = () => {
    const navigate = useNavigate();
    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);
    const heroes = getHeroesByName(q);
    const showSearch = (q.length === 0);
    const showError = (q.length > 0 && heroes.length === 0);
    const { searchText, onInputChange, onResetForm } = useForm({ searchText: '' });

    const onSubmitSearch = (event) => {
        event.preventDefault();
        navigate(`?q=${searchText}`);
        onResetForm();
    };

    return (
        <>
            <h1 className="mt-2">Busca tu héroe</h1>
            <hr />
            <div className="row mt-5">
                <div className="col-5">
                    <h4>Buscando</h4>
                    <hr />

                    <form className="d-flex" onSubmit={onSubmitSearch}>
                        <input
                            className="form-control me-2"
                            type="search"
                            placeholder="Search"
                            aria-label="Search"
                            name="searchText"
                            value={searchText}
                            onChange={onInputChange}
                        />
                        <button className="btn btn-outline-primary" type="submit">
                            Buscar
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Resultados:</h4>
                    <hr />
                    <div
                        aria-label="alert"
                        className="alert alert-primary animate__animated animate__fadeIn"
                        style={{ display: showSearch ? '' : 'none' }}
                    >
                        Busca tu héroe
                    </div>
                    <div
                        aria-label="error"
                        className="alert alert-danger animate__animated animate__fadeIn"
                        style={{ display: showError ? '' : 'none' }}
                    >
                        No hay resultados para <b>{q}</b>
                    </div>
                    {
                        heroes.map(hero => (
                            <HeroCard key={hero.id} {...hero} />
                        ))
                    }
                </div>
            </div>
        </>
    )
}
