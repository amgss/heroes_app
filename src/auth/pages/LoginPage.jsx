import { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../context/AuthContext';

export const LoginPage = () => {
    const { login } = useContext(AuthContext);
    const navigate = useNavigate();

    const onLogin = () => {
        const lastPath = localStorage.getItem('lastPath') || '/';
        login('antonio.martin');
        console.log('login in...');
        navigate(lastPath, {
            replace: true
        });
    }

    return (
        <div className="container mt-5">
            <h1>LoginPage</h1>
            <hr />

            <button
                className="btn btn-outline-primary"
                onClick={onLogin}
            >
                Login
            </button>
        </div>
    )
}
