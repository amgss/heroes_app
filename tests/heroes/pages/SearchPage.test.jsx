import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { AuthContext } from "../../../src/auth/context/AuthContext";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

describe('Pruebas sobre <SearchPage/>', () => {
    const contextValue = {
        logged: true,
        user: { id: 1, name: 'Usuario' },
        login: jest.fn(),
        logout: jest.fn()
    };

    beforeEach(() => jest.clearAllMocks());

    test('Debe mostrarse correctamente con valores por defecto', () => {
        const {container} = render(
            <MemoryRouter initialEntries={['/search']}>
                <AuthContext.Provider value={contextValue}>
                    <SearchPage />
                </AuthContext.Provider>
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
    });

    test('Debe mostrar a Batman y el input vacÃ­o', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <AuthContext.Provider value={contextValue}>
                    <SearchPage />
                </AuthContext.Provider>
            </MemoryRouter>
        );

        const input = screen.getByRole('searchbox');
        expect(input.value).toBe('');
        expect(screen.getByText('Batman')).toMatchSnapshot();

        const img = screen.getByRole('img');
        expect(img.src).toContain('dc-batman');

        const alert = screen.getByLabelText('alert');
        expect(alert.style.display).toBe('none');

        screen.debug();
    });
});