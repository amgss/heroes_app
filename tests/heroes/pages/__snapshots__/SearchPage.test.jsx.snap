// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <SearchPage/> Debe mostrar a Batman y el input vacío 1`] = `
<h5
  class="card-title"
>
  Batman
</h5>
`;

exports[`Pruebas sobre <SearchPage/> Debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1
    class="mt-2"
  >
    Busca tu héroe
  </h1>
  <hr />
  <div
    class="row mt-5"
  >
    <div
      class="col-5"
    >
      <h4>
        Buscando
      </h4>
      <hr />
      <form
        class="d-flex"
      >
        <input
          aria-label="Search"
          class="form-control me-2"
          name="searchText"
          placeholder="Search"
          type="search"
          value=""
        />
        <button
          class="btn btn-outline-primary"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Resultados:
      </h4>
      <hr />
      <div
        aria-label="alert"
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Busca tu héroe
      </div>
      <div
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hay resultados para 
        <b />
      </div>
    </div>
  </div>
</div>
`;
